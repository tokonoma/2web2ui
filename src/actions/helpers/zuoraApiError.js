import _ from 'lodash';

// TODO: Return an Error object as if we were to create it with class and extends Error
function ZuoraApiError(error) {
  const message = _.get(
    error.response,
    'data.reasons[0].message',
    'An error occurred while contacting the billing service',
  );

  this.name = 'ZuoraApiError';
  this.stack = error.stack; // must manually assign prototype value
  this.message = stripTags(message); // some messages include html tags

  // Intentionally assigning additional properties
  Object.assign(this, error);
}

ZuoraApiError.prototype = Object.create(Error.prototype);

export default ZuoraApiError;
