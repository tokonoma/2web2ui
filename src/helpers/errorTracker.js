import _ from 'lodash';
import Raven from 'raven-js';
import createRavenMiddleware from 'raven-for-redux';
import bowser from 'bowser';

/**
 * List of breadcrumbs to ignore to reduce noise
 * @note Sentry has a 100 breadcrumb per event/error maximum
 */
const BLACKLIST = new Set([
  '@@redux/INIT',
  '@@redux-form/BLUR',
  '@@redux-form/CHANGE',
  '@@redux-form/DESTROY',
  '@@redux-form/FOCUS',
  '@@redux-form/REGISTER_FIELD',
  '@@redux-form/TOUCH',
  '@@redux-form/UNREGISTER_FIELD',
  '@@redux-form/UPDATE_SYNC_ERRORS',
  '@@redux-form/INITIALIZE',
  '@@redux-form/STOP_SUBMIT',
  '@@redux-form/SET_SUBMIT_FAILED',
]);

/**
 * Filter out blacklisted breadcrumbs
 *
 * @param {object} crumb
 * @param {string} crumb.category - console, http, redux-action, xhr, etc.
 * @param {string} crumb.message - error message, redux action type, etc.
 * @param {number} crumb.timestamp - created at
 * @see https://docs.sentry.io/clients/javascript/config
 */
export function breadcrumbCallback(crumb) {
  if (BLACKLIST.has(crumb.message)) {
    return false;
  }

  return crumb;
}

/**
 * Remove sensitive tokens from URLs
 * @param {String} url
 */
function filterURL(url) {
  const filters = [
    { re: /\/reset-password\/.*$/, replacement: '/reset-password/[FILTERED]' },
    {
      re: /\/account\/email-verification\/.*$/,
      replacement: '/account/email-verification/[FILTERED]',
    },
  ];

  if (!url) {
    return url;
  }

  // url.replace(filters[0].re, filters[0].replacement)
  //    .replace(filters[1].re, filters[1].replacement) ...
  return filters.reduce((filtered, { re, replacement }) => filtered.replace(re, replacement), url);
}

/**
 * Sanitise any URL in a given field of an object.
 * @param {Object} obj - containing key
 * @param {String} key - key to sanitise
 */
function filterUrlFromKeyIn(obj, key) {
  if (!obj) {
    return obj;
  }
  return {
    ...obj,
    [key]: filterURL(obj[key]),
  };
}

/**
 * Sanitise the browser request URL
 * @param {Object} request - Sentry request object
 */
function filterRequest(request) {
  if (!request) {
    return request;
  }

  const urlFiltered = filterUrlFromKeyIn(request, 'url');
  const headers = filterUrlFromKeyIn(request.headers, 'Referer');

  return {
    ...urlFiltered,
    headers,
  };
}

// Closure to safely enrich events with data from Redux store
export function getEnricherOrDieTryin(store, currentWindow) {
  return function enrich(data) {
    const { currentUser } = store.getState();
    const user = _.pick(currentUser, ['access_level', 'customer', 'username']);
    const fromOurBundle = isErrorFromOurBundle(data);
    const apiError = isApiError(data);
    const chunkFailure = isChunkFailure(data);
    const request = filterRequest(data.request);
    const userAgent = _.get(currentWindow, 'navigator.userAgent');

    /*global SUPPORTED_BROWSERS*/
    // refer to docs/browser-support-sentry-issue.md for more info
    const isSupportedBrowser = Boolean(bowser.getParser(userAgent).satisfies(SUPPORTED_BROWSERS));

    return {
      ...data,
      extra: {
        // extra data that doesn't need to be searchable
        ...data.extra,
        isApiError: apiError,
        isChunkFailure: chunkFailure,
        isSupportedBrowser,
      },
      request,
      level:
        !fromOurBundle || apiError || chunkFailure || !isSupportedBrowser ? 'warning' : 'error',
      tags: {
        // all tags can be easily searched and sent in Slack notifications
        ...data.tags,
        customer: _.get(user, 'customer'),
        // This <html> property should be set by us and updated when page is translated
        documentLanguage: _.get(currentWindow, 'document.documentElement.lang', 'unknown'),
        navigatorLanguage: _.get(currentWindow, 'navigator.language', 'unknown'),
        source: fromOurBundle ? '2web2ui' : 'unknown',
      },
      user,
    };
  };
}

// Check if error event was thrown from our bundle or from something else (i.e. browser extension)
export function isErrorFromOurBundle(data) {
  // The local environment match is looser to allow for hot module replacement (i.e. http://app.sparkpost.test/4.a0803f8355f692de1382.hot-update.js)
  const looksLikeOurBundle = /sparkpost\.test|sparkpost\.com\/static\/js\//;
  // There should never be multiple exception values
  let frames = _.get(data, 'exception.values[0].stacktrace.frames', []);
  const firstFunction = _.get(frames, '[0].function');

  // A Sentry function may be included in the stacktrace and needs to be ignored (i.e. HTMLDocument.wrapped or wrapped)
  if (/wrapped/.test(firstFunction)) {
    frames = frames.slice(1); // safely reassign frames without Sentry function
  }

  // if any frame is from our bundle
  return Boolean(frames.find(({ filename }) => looksLikeOurBundle.test(filename)));
}

export function isApiError(data) {
  const looksLikeApiErrorType = /SparkpostApiError|ZuoraApiError/;
  const type = _.get(data, 'exception.values[0].type');

  return looksLikeApiErrorType.test(type);
}

export function isChunkFailure(data) {
  const looksLikeChunkFailure = /Loading chunk/;
  const value = _.get(data, 'exception.values[0].value');

  return looksLikeChunkFailure.test(value);
}

// The purpose of this helper is to provide a common interface for reporting errors
// with the expectation that the current service will change in the future.
class ErrorTracker {
  /**
   * The service must be configured before it can be used
   * @param {object} store - the Redux store for additional context
   * @param {object}
   */
  install(config, store) {
    const { release, sentry, tenantId } = config;

    // Silently ignore installation if Sentry configuration is not provided
    if (!sentry) {
      return;
    }

    const dsn = `https://${sentry.publicKey}@sentry.io/${sentry.projectId}`;
    const options = {
      breadcrumbCallback,
      dataCallback: getEnricherOrDieTryin(store, window),
      release,
      tags: { tenant: tenantId },
    };

    Raven.config(dsn, options).install();
  }

  // Record redux actions as breadcrumbs
  get middleware() {
    return createRavenMiddleware(Raven);
  }

  /**
   * Report an error
   *
   * @param {string} logger - simple description of where the error was caught
   * @param {Error} error
   * @example
   *   try {
   *     throw new Error('Oh no, save me!');
   *   } catch(error) {
   *     ErrorTracker.report('where-am-i', error);
   *   }
   */
  report(logger, error, extra = {}) {
    // Silently ignore if Sentry is not setup
    if (!Raven.isSetup()) {
      return;
    }
    Raven.captureException(error, { logger, extra });
  }

  addRequestContextAndThrow(reduxActionType, response = {}, error) {
    if (!Raven.isSetup()) {
      throw error;
    }
    let zuoraErrorCode = _.get(response, 'data.reasons', [])
      .map(reason => reason.code)
      .join();

    Raven.context(
      {
        tags: {
          reduxActionType,
          httpResponseStatus: response.status || 0,
          zuoraErrorCode: zuoraErrorCode,
        },
      },
      () => {
        throw error;
      },
    );
  }
}

// Export a constructed instance to avoid the need to call `new` everywhere
export default new ErrorTracker();
