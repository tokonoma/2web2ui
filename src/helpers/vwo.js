export function trackCustomConversionGoal(...args) {
  window._vis_opt_queue = window._vis_opt_queue || [];
  window._vis_opt_queue.push(function() {
    window._vis_opt_register_conversion(...args);
  });
}

export function hasVWOflagSet(option) {
  return Boolean(window[option]);
}
