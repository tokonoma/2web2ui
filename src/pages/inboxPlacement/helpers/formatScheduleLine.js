import React from 'react';
import moment from 'moment';
import { STATUS, DURATION_HOURS } from '../constants/test';
import { FORMATS } from 'src/constants';
import { InlineItems } from 'src/components/structure';
import { Bold } from 'src/components/text';

const formatDateTime = start_time => moment(start_time).format(FORMATS.LONG_DATETIME_ALT);
const getHoursRemaining = start_time =>
  moment(start_time)
    .add(DURATION_HOURS, 'hours')
    .diff(moment(), 'hours');

export default function formatScheduleLine(status, start_time, end_time) {
  const formatted_start_time = formatDateTime(start_time);

  switch (status) {
    case STATUS.RUNNING: {
      return (
        <>
          <Bold>Sent</Bold> {formatted_start_time} and {getHoursRemaining(start_time)} hours
          remaining on test
        </>
      );
    }
    case STATUS.STOPPED: {
      return (
        <InlineItems compact>
          <span>
            <Bold>Sent</Bold> {formatted_start_time}
          </span>
          <span>
            <Bold>Stopped</Bold> {formatDateTime(end_time)}
          </span>
        </InlineItems>
      );
    }
    case STATUS.COMPLETED: {
      return (
        <InlineItems compact>
          <span>
            <Bold>Sent</Bold> {formatted_start_time}
          </span>
          <span>
            <Bold>Completed</Bold> {formatDateTime(end_time)}
          </span>
        </InlineItems>
      );
    }
    default: {
      return (
        <>
          <Bold>Sent</Bold> {formatted_start_time} and test status unknown
        </>
      );
    }
  }
}
