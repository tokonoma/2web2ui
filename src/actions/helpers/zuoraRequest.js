import requestHelperFactory from './requestHelperFactory';
import _ from 'lodash';
import { zuora as zuoraAxios } from 'src/helpers/axiosInstances';
import ErrorTracker from 'src/helpers/errorTracker';
import { showAlert } from 'src/actions/globalAlert';
import { stripTags } from 'src/helpers/string';

export default requestHelperFactory({
  request: zuoraAxios,
  onSuccess: ({ types, response, dispatch, meta }) => {
    if (!response.data.success) {
      try {
        const message = _.get(
          response,
          'data.reasons[0].message',
          'An error occurred while contacting the billing service',
        );
        const sanitizedMessage = stripTags(message); // some messages include html tags
        const error = new Error(sanitizedMessage);
        error.name = 'ZuoraApiError';
        error.response = response;

        dispatch({
          type: types.FAIL,
          payload: { error, message: sanitizedMessage, response },
          meta,
        });

        // auto alert all errors
        dispatch(showAlert({ type: 'error', message: sanitizedMessage }));

        // TODO: 'return' err once we unchain all actions
        ErrorTracker.addRequestContextAndThrow(types.FAIL, response, error);
      } catch (err) {
        ErrorTracker.report('zuora-request-error', err);
      }
    }

    dispatch({
      type: types.SUCCESS,
      payload: response,
      meta,
    });

    return meta.onSuccess ? dispatch(meta.onSuccess({ results: response })) : response;
  },
  onFail: ({ types, err, dispatch }) => {
    dispatch(showAlert({ type: 'error', message: 'An error occurred while contacting zuora' }));
    //report 4xx and 5xx errors to sentry
    ErrorTracker.addRequestContextAndThrow(types.FAIL, err.response, err);
  },
});
