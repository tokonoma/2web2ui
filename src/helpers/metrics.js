import moment from 'moment';
import _ from 'lodash';
import { list as METRICS_LIST } from 'src/config/metrics';
import config from 'src/config';
import { HIBANA_METRICS_COLORS } from 'src/constants/index';
import { getRelativeDates } from 'src/helpers/date';
import { safeDivide, safeRate } from './math';

const {
  metricsPrecisionMap: precisionMap,
  metricsRollupPrecisionMap: rollupPrecisionMap,
  apiDateFormat,
  chartColors = [],
} = config;
const indexedPrecisions = _.keyBy(precisionMap, 'value');
const FILTER_KEY_MAP = {
  'Recipient Domain': 'domains',
  Campaign: 'campaigns',
  Template: 'templates',
  'Sending IP': 'sending_ips',
  'IP Pool': 'ip_pools',
  Subaccount: 'subaccounts',
  'Sending Domain': 'sending_domains',
};

const DELIMITERS = ',;:+~`!@#$%^*()-={}[]"\'<>?./|\\'.split('');

export function getQueryFromOptions({
  from,
  to,
  timezone,
  precision,
  metrics,
  filters = [],
  match = '',
  limit,
}) {
  from = moment(from);
  to = moment(to);

  const apiMetricsKeys = getKeysFromMetrics(metrics);
  const delimiter = getDelimiter(filters);
  const options = {
    metrics: apiMetricsKeys.join(delimiter),
    from: from.format(apiDateFormat),
    to: to.format(apiDateFormat),
    delimiter,
    ...getFilterSets(filters, delimiter),
    timezone,
    precision,
  };
  if (match.length > 0) {
    options.match = match;
  }
  if (limit) {
    options.limit = limit;
  }
  return options;
}

export function pushToKey(obj, key, value) {
  const updated = { [key]: [], ...obj };
  updated[key].push(value);
  return updated;
}

export function getFilterSets(filters = [], delimiter) {
  const hash = filters.reduce(
    (result, { type, value, id }) =>
      pushToKey(result, FILTER_KEY_MAP[type], type === 'Subaccount' ? id : value),
    {},
  );
  return _.mapValues(hash, v => v.join(delimiter));
}

/**
 * Get an array of unique chars based on all filters and
 * find all delimiters that don't appear in that list, then
 * return the first one to use as the delimiter
 *
 * @param {array} filters
 */
export function getDelimiter(filters = []) {
  const uniques = _.uniq(filters.map(f => f.value).join(''));
  return _.difference(DELIMITERS, uniques).shift();
}

/**
 * Calculates the difference between 2 moment dates
 * and returns the closest precision value
 *
 */
export function getPrecision(from, to = moment()) {
  const diff = to.diff(from, 'minutes');
  return precisionMap.find(({ time }) => diff <= time).value;
}

/**
 * Calculates the precision value for metrics Rollup. If the precision given
 * is still within range, do not change precision. If the possible precision options
 * do not include the current precision, get the recommended precision.
 */
export function getRollupPrecision({ from, to = moment(), precision }) {
  if (!precision) {
    return undefined;
  }
  const precisionOptions = getPrecisionOptions(moment(from), moment(to));
  const precisionOptionsValues = precisionOptions.map(({ value }) => value);
  if (precisionOptionsValues.includes(precision)) {
    return precision;
  }
  return getRecommendedRollupPrecision(from, to);
}
export function getRecommendedRollupPrecision(from, to = moment()) {
  const diff = moment(to).diff(moment(from), 'minutes');
  return rollupPrecisionMap.find(({ recommended }) => diff <= recommended).value;
}

/**
 * Creates an array of possible precision options for a time span
 */
export function getPrecisionOptions(from, to = moment()) {
  const diff = to.diff(from, 'minutes');
  return rollupPrecisionMap
    .filter(({ min, max }) => diff >= min && diff <= max)
    .map(({ value }) => ({ value, label: _.startCase(_.words(value).join(' ')) }));
}

export function getMomentPrecisionByDate(from, to = moment()) {
  const diff = to.diff(from, 'minutes');
  if (diff <= 60 * 24) {
    return 'minutes';
  }
  return diff <= 60 * 24 * 2 ? 'hours' : 'days';
}

export function getMomentPrecisionByPrecision(precision) {
  if (precision.includes('min')) {
    return 'minutes';
  }
  if (['day', 'week', 'month'].includes(precision)) {
    return 'day';
  }
  return 'hour';
}

export function getPrecisionType(precision) {
  return indexedPrecisions[precision].time <= 60 * 24 * 2 ? 'hour' : precision;
}

// We are forced to use UTC for any precision greater or equal to 'day'
const FORCED_UTC_ROLLUP_PRECISIONS = ['day', 'week', 'month'];
export function isForcedUTCRollupPrecision(precision) {
  return FORCED_UTC_ROLLUP_PRECISIONS.includes(precision);
}

/**
 * Round 'from' and 'to' to nearest precision
 *
 * @param fromInput
 * @param toInput
 * @param precision,
 * @return {{to: *|moment.Moment, from: *|moment.Moment}}
 */
export function roundBoundaries({
  from: fromInput,
  to: toInput,
  now = moment(),
  precision: defaultPrecision,
}) {
  const from = moment(fromInput);
  const to = moment(toInput);

  const precision = defaultPrecision || getPrecision(from, to);
  const momentPrecision = getMomentPrecisionByPrecision(precision);
  // extract rounding interval from precision query param value
  const roundInt = momentPrecision === 'minutes' ? parseInt(precision.replace('min', '')) : 1;

  floorMoment(from, roundInt, momentPrecision);

  // if we're only at a minute precision, don't round up to the next minute
  const isToSameAsNow = Math.abs(moment(now).diff(to, 'minutes')) < 1;

  if (precision !== '1min' && !isToSameAsNow) {
    ceilMoment(to, roundInt, momentPrecision);
  }

  return { to, from };
}

/**
 * Rounds down moment to precision and interval
 * @param time
 * @param roundInt
 * @param precision
 */
function floorMoment(time, roundInt = 1, precision = 'minutes') {
  const value = time.get(precision);
  const roundedValue = Math.floor(value / roundInt) * roundInt;
  time.set(precision, roundedValue).startOf(precision);
}

/**
 * Rounds up moment to precision and interval
 * @param time
 * @param roundInt
 * @param precision
 */
function ceilMoment(time, roundInt = 1, precision = 'minutes') {
  const value = time.get(precision);
  const roundedValue = Math.ceil(value / roundInt) * roundInt;
  time.set(precision, roundedValue).endOf(precision);
}

/**
 * Returns verified from and to dates; throws an error if date range is invalid - catch this to reset to last state
 *
 * @param from
 * @param to
 * @param now
 * @param roundToPrecision
 * @return {*}
 */
export function getValidDateRange({
  from,
  to,
  now = moment(),
  roundToPrecision,
  preventFuture,
  precision,
}) {
  // If we're not rounding, check to see if we want to prevent future dates. if not, we're good.
  const nonRoundCondition = roundToPrecision || precision || !preventFuture || to.isBefore(now);
  const validDates = _.every(
    _.map([from, to, now], date => moment.isMoment(date) && date.isValid()),
  );

  if (validDates && from.isBefore(to) && nonRoundCondition) {
    if (!roundToPrecision) {
      return { from, to };
    }
    // Use the user's rounded 'to' input if it's less than or equal to 'now' rounded up to the nearest precision,
    // otherwise use the valid range and precision of floor(from) to ceil(now).
    // This is necessary because the precision could change between the user's invalid range, and a valid range.
    const { from: roundedFromNow, to: roundedNow } = roundBoundaries({ from, to: now, precision });
    const { from: roundedFrom, to: roundedTo } = roundBoundaries({ from, to, precision });

    if (roundedTo.isSameOrBefore(roundedNow)) {
      from = roundedFrom;
      to = roundedTo;
    } else {
      from = roundedFromNow;
      to = roundedNow;
    }

    return { from, to };
  }

  throw new Error('Invalid date range selected');
}

export function _getMetricsFromKeys(keys = [], isHibanaEnabled) {
  const metricsColors = isHibanaEnabled ? HIBANA_METRICS_COLORS : chartColors;
  return keys.map((metric, i) => {
    const found = METRICS_LIST.find(M => M.key === metric || M.key === metric.key);

    if (!found) {
      throw new Error(`Cannot find metric: ${JSON.stringify(metric)}`);
    }

    return { ...found, name: found.key, stroke: metricsColors[i % metricsColors.length] };
  });
}

export const getMetricsFromKeys = _.memoize(
  _getMetricsFromKeys,
  (keys = [], isHibanaEnabled) => `${keys.join('')}${isHibanaEnabled ? 'hibana' : 'og'}`,
);

export function getKeysFromMetrics(metrics = []) {
  const flattened = _.flatMap(metrics, ({ key, computeKeys }) => (computeKeys ? computeKeys : key));
  return _.uniq(flattened);
}

export function computeKeysForItem(metrics = []) {
  return item =>
    metrics.reduce((acc, metric) => {
      if (metric.compute) {
        acc[metric.key] = metric.compute(acc, metric.computeKeys) || 0;
      }
      return acc;
    }, item);
}

/**
 * Transforms API result into chart-ready data in 2 steps:
 * 1. compute any necessary computed metrics
 * 2. arrange metrics into groups sorted by unit/measure
 *
 * @param {Array} metrics - list of currently selected metrics objects from config
 * @param {Array} data - results from the metrics API
 */
export function transformData(data = [], metrics = []) {
  return data.map(computeKeysForItem(metrics));
}

// Extract from, to, filters (campaign, template, ...) and any other included update fields
// into a set of common options for metrics queries.
export function buildCommonOptions(reportOptions, updates = {}) {
  return {
    ...reportOptions,
    ...updates,
    ...getRelativeDates(reportOptions.relativeRange),
    ...getRelativeDates(updates.relativeRange),
  };
}

export function average(item, keys = []) {
  return safeDivide(item[keys[0]], item[keys[1]]);
}

export function rate(item, keys = []) {
  return safeRate(item[keys[0]], item[keys[1]]);
}
