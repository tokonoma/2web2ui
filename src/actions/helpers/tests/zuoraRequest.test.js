import zuoraRequest from '../zuoraRequest';
import { createMockStore } from 'src/__testHelpers__/mockStore';
import * as axiosMocks from 'src/helpers/axiosInstances';
import ErrorTracker from 'src/helpers/errorTracker';

jest.mock('src/helpers/axiosInstances');
jest.mock('src/helpers/errorTracker');

ErrorTracker.addRequestContextAndThrow = jest.fn((type, response, err) => {
  throw err;
});

describe('Helper: Zuora API Request', () => {
  let mockStore;
  let action;
  let expectedResponse;

  beforeEach(() => {
    mockStore = createMockStore({});
    action = { type: 'ZUORA_REQUEST', meta: {} };
  });

  it('should make a successful call', async () => {
    expectedResponse = { data: { success: true } };
    axiosMocks.zuora.mockImplementation(() => Promise.resolve(expectedResponse));
    const results = await mockStore.dispatch(zuoraRequest(action));
    expect(results).toBe(expectedResponse);
    expect(mockStore.getActions()).toMatchSnapshot();
  });

  it('should handle a zuora failure', async () => {
    axiosMocks.zuora.mockImplementation(() =>
      Promise.resolve({
        data: {
          success: false,
          reasons: [{ message: 'The credit card is bad and wrong' }],
        },
      }),
    );

    await mockStore.dispatch(zuoraRequest(action));
    expect(mockStore.getActions()).toMatchSnapshot();
  });
});
