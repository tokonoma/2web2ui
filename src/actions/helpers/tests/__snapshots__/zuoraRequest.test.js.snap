// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper: Zuora API Request should handle a zuora failure 1`] = `
Array [
  Object {
    "meta": Object {},
    "type": "ZUORA_REQUEST_PENDING",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "error": [ZuoraApiError: The credit card is bad and wrong],
      "message": "The credit card is bad and wrong",
      "response": Object {
        "data": Object {
          "reasons": Array [
            Object {
              "message": "The credit card is bad and wrong",
            },
          ],
          "success": false,
        },
      },
    },
    "type": "ZUORA_REQUEST_FAIL",
  },
  Object {
    "payload": Object {
      "message": "The credit card is bad and wrong",
      "type": "error",
    },
    "type": "SHOW_GLOBAL_ALERT",
  },
]
`;

exports[`Helper: Zuora API Request should make a successful call 1`] = `
Array [
  Object {
    "meta": Object {},
    "type": "ZUORA_REQUEST_PENDING",
  },
  Object {
    "meta": Object {},
    "payload": Object {
      "data": Object {
        "success": true,
      },
    },
    "type": "ZUORA_REQUEST_SUCCESS",
  },
]
`;
